Command display overhaul

Uzair5162 opened the pull request:
Mainly a command display rework, with related changes

#### Blank param indexing:
- refactors `logic::index_parameters_for_display()` to only index blank params in a command, and not style them with `<bold><italics>` tags
  - styling now happens in the cli crate (as it should) and allows for re-use of `logic::index_parameters_for_display()` by the ui
- the _indexed_ command is displayed under the generated command box
<img width="677" alt="image" src="https://github.com/user-attachments/assets/9b3e62c6-e431-4ca5-9386-5aa4c5ef2ec3" />

#### Saved and edit command:
- the saved command is displayed under the generated command box (see above image) with indexed params
- on edit, this turns editable and indexes disappear (showing the original command)
<img width="672" alt="image" src="https://github.com/user-attachments/assets/20b4b3b3-c862-437d-af28-9fde6181cd72" />

#### Note:
- note is displayed below the saved command, with "+ Add a note" button if none
  - on click, the button turns into the text field for the note
<img width="320" alt="image" src="https://github.com/user-attachments/assets/2a87d07c-5324-4ef7-9f9d-c793270d0662" />
<img width="320" alt="image" src="https://github.com/user-attachments/assets/3c665bd2-c492-4900-b329-b8f7fce792d4" />

#### Param viewer
- styles the param box and labels blank params with their index
<img width="675" alt="image" src="https://github.com/user-attachments/assets/3a10facb-7bdf-4f0b-b608-156d4e07163d" />

#### Misc.
- un-hardcodes the scrollable section height on the viewer (everything below saved command)
- separates some things into components in command-display folder
- commented tag and favourite edit for now, need to be styled in future work

#### Before
<img width="1552" alt="image" src="https://github.com/user-attachments/assets/6646d9d5-62aa-4a6e-8047-34dc63b35694" />

#### After
<img width="1552" alt="image" src="https://github.com/user-attachments/assets/84a8a5d8-7b0c-4cd4-8616-5ae6b85ff7c5" />

Uzair5162 committed:
refactor command indexing
- moves styling (bold and italicizing) to cli crate
- keeps numbering in logic crate

Uzair5162 committed:
add tauri command indexing handler

Uzair5162 committed:
style command display and index params on ui

Uzair5162 committed:
clippy

Uzair5162 committed:
fmt

danyal002 commented:
A few comments:

* Why did you disable editing of parameters? We've been in agreement since before that we will have that box editable.  The reason for this was that it could be the case that the user likes what was randomly generated, but doesn't want all of it, but only part of it. As always, if the user needs to use notepad or another text editor to fix what we generate, then we've failed. Please enable that again.


* Honestly, I don't like how the parameters section looks now. There's just too much going on. I'm not a big fan of the font. It's hard to read (see first screenshot). I would prefer that we revert to the previous implementation where the String min, max was a label on the textfield itself. It looked a lot cleaner. In the current implementation, I'm having to trace my eyes from the String min, max on the left to the field itself on the right.
![Screenshot_2025-03-05_at_15 18 07](https://github.com/user-attachments/assets/d4940ae5-d75c-4937-a0dc-c7ce0fe5569e)
* I like the "Fill in" placeholder in the blank parameter textfield. Upon reverting to the previous implementation, we can reuse the Blank @{1} text to be the label.

* I don't like how we've removed the Note text box and its heading. It looks like it's not editable when it is. Again, I think we should switch back to the previous implementation.

* I think the order can be changed. The note can be moved all the way down. We can dynamically calculate the height of the parameters scrollpane so that the note section doesn't disappear from view.

* I think there should be more padding between the centre divider line and where the Note and Parameters text are (see second screenshot).
![Screenshot_2025-03-05_at_15 19 45](https://github.com/user-attachments/assets/5fa1245e-defd-4113-a7e0-1de4f697d326)

Uzair5162 commented:
> A few comments:
> 
> * Why did you disable editing of parameters? We've been in agreement since before that we will have that box editable.  The reason for this was that it could be the case that the user likes what was randomly generated, but doesn't want all of it, but only part of it. As always, if the user needs to use notepad or another text editor to fix what we generate, then we've failed. Please enable that again.

Parameters were never editable.

> * Honestly, I don't like how the parameters section looks now. There's just too much going on. I'm not a big fan of the font. It's hard to read (see first screenshot). I would prefer that we revert to the previous implementation where the String min, max was a label on the textfield itself. It looked a lot cleaner. In the current implementation, I'm having to trace my eyes from the String min, max on the left to the field itself on the right.
>   ![Screenshot_2025-03-05_at_15 18 07](https://private-user-images.githubusercontent.com/50880807/419636557-d4940ae5-d75c-4937-a0dc-c7ce0fe5569e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyOTIzNjQsIm5iZiI6MTc0MTI5MjA2NCwicGF0aCI6Ii81MDg4MDgwNy80MTk2MzY1NTctZDQ5NDBhZTUtZDc1Yy00OTM3LWEwZGMtYzdjZTBmZTU1NjllLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA2VDIwMTQyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlOGYyMTExZTFjMzQzODA5NmFmODM3ZTI2Njg0YTdjZTcxN2JlOTI2OTM2ODFjOGM5NDUyNmU4YzAxZDk3YjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Xf2jvn81js5IKVq4aAbz_D_a1INBMNBKW1CfOZBAwH0)
> * I like the "Fill in" placeholder in the blank parameter textfield. Upon reverting to the previous implementation, we can reuse the Blank @{1} text to be the label.

I'll change this and generated box font if you have other mono font suggestions. Re: being a label on the textfield itself, I don't like how everything was stacked vertically in the prev ui. I think it looks better side-by-side since it looks less empty. We can discuss alternatives with horizontal spacing and other ppls thoughts.

> * I don't like how we've removed the Note text box and its heading. It looks like it's not editable when it is. Again, I think we should switch back to the previous implementation.

Not sure what you mean. The note box was never editable even though we displayed it in a textarea. It only became editable after clicking edit and that is still the case. Can also make it so that the cursor focuses on the textbox on edit if that helps.

> * I think the order can be changed. The note can be moved all the way down. We can dynamically calculate the height of the parameters scrollpane so that the note section doesn't disappear from view.

Agree.

> * I think there should be more padding between the centre divider line and where the Note and Parameters text are (see second screenshot).
>   ![Screenshot_2025-03-05_at_15 19 45](https://private-user-images.githubusercontent.com/50880807/419636641-5fa1245e-defd-4113-a7e0-1de4f697d326.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyOTIzNjQsIm5iZiI6MTc0MTI5MjA2NCwicGF0aCI6Ii81MDg4MDgwNy80MTk2MzY2NDEtNWZhMTI0NWUtZGVmZC00MTEzLWE3ZTAtMWRlNGY2OTdkMzI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA2VDIwMTQyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2YzIzOGY2ZTU5ZDU0MzRjNzk4MDM1ZjRmMTRiYjEwYjg3YjdjNDI5OWI1ZDkwNmQzOWRiYzU2NmRmNDFkYzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bXvppyIUf8UP_HC-NZqXx5uqzz8sS05CJr2mvRJbtlQ)

Will see how this looks.

Pop0097 commented:
> Honestly, I don't like how the parameters section looks now. There's just too much going on. I'm not a big fan of the font. It's hard to read (see first screenshot). I would prefer that we revert to the previous implementation where the String min, max was a label on the textfield itself. It looked a lot cleaner. In the current implementation, I'm having to trace my eyes from the String min, max on the left to the field itself on the right.

I personally like the new layout better, but I do agree that the font could be changed.

Pop0097 commented:
Some other comments:
- I think when editing a command it isn't very obvious that the user should be clicking the following to change the command (see first image below). I am not too sure what the best path forward for this would be. One idea that I have is removing the command preview box when editing and converting the command to a text box similar to the note (see second image below):

<img width="829" alt="image" src="https://github.com/user-attachments/assets/4b88e71d-9b27-457b-838c-462e0eb8f953" />

<img width="833" alt="image" src="https://github.com/user-attachments/assets/95a9579c-53c1-4947-9961-d92b96166042" />

- When parameters become very long, they get truncated 

<img width="826" alt="Screenshot 2025-03-07 at 13 10 43" src="https://github.com/user-attachments/assets/afd00d25-053b-4746-8fef-63eae5c43ff3" />

Another nit:
- Is there a way that a user can cancel editing? When editing I only see "Save" and "Delete Command" as an option

Uzair5162 committed:
change mono font to roboto

Uzair5162 committed:
rework note and generated cmd edit

Uzair5162 committed:
rework param viewer styling

Uzair5162 committed:
remove disabled prop from usecommandbox

Uzair5162 commented:
- changed font in params and display/saved command boxes to roboto mono
- hides generated command box on edit
- adds Note heading and moved it below params
- brings param label and display/input closer together and makes overflow scroll
<img width="1552" alt="image" src="https://github.com/user-attachments/assets/ca3757f7-4c01-48c0-9c67-96c0bf5aefd9" />

danyal002 commented on ui/src-tauri/src/lib.rs:
```diff
@@ -155,6 +156,19 @@ fn replace_parameters(
 
+#[tauri::command]
+fn index_blank_parameters(command: String, state: State<Ui>) -> Result<String, UiError> {
+    if let Ok(logic) = state.logic.write() {
```
You might be able to get away with using a `read` lock here since you're not modifying anything.

Uzair5162 committed:
use read lock
